nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchHttp.java

```diff
@@ -0,0 +1,330 @@
+        // Identifier Attribute property is empty, the Index Operation must evaluate to "index".
+        String idAttribute = validationContext.getProperty(ID_ATTRIBUTE).getValue();
+        String indexOp = validationContext.getProperty(INDEX_OP).getValue();
+
+        if(StringUtils.isEmpty(idAttribute)) {
+            if(StringUtils.isEmpty(indexOp)) {
+                problems.add(new ValidationResult.Builder()
+                        .valid(false)
+                        .subject(INDEX_OP.getDisplayName())
+                        .explanation("If Identifier Attribute is not set, Index Operation must evaluate to \"index\"")
+                        .build());
+            }
+        }
+        return problems;
+    }
+
+    @OnScheduled
+    public void setup(ProcessContext context) {
+        super.setup(context);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        final String id_attribute = context.getProperty(ID_ATTRIBUTE).getValue();
+        final Charset charset = Charset.forName(context.getProperty(CHARSET).getValue());
+
+        final List<FlowFile> flowFiles = session.get(batchSize);
+        if (flowFiles.isEmpty()) {
+            return;
+        }
+
+        OkHttpClient okHttpClient = getClient();
+        final ComponentLog logger = getLogger();
+
+        // Keep track of the list of flow files that need to be transferred. As they are transferred, remove them from the list.
+        List<FlowFile> flowFilesToTransfer = new LinkedList<>(flowFiles);
+        try {
+            final StringBuilder sb = new StringBuilder();
+            final String baseUrl = trimToEmpty(context.getProperty(ES_URL).getValue());
+            final URL url = new URL((baseUrl.endsWith("/") ? baseUrl : baseUrl + "/") + "_bulk");
+
+            for (FlowFile file : flowFiles) {
+                final String index = context.getProperty(INDEX).evaluateAttributeExpressions(file).getValue();
+                if (StringUtils.isEmpty(index)) {
+                    logger.error("No value for index in for {}, transferring to failure", new Object[]{id_attribute, file});
+                    flowFilesToTransfer.remove(file);
+                    session.transfer(file, REL_FAILURE);
+                    continue;
+                }
+                final String docType = context.getProperty(TYPE).evaluateAttributeExpressions(file).getValue();
+                String indexOp = context.getProperty(INDEX_OP).evaluateAttributeExpressions(file).getValue();
+                if (StringUtils.isEmpty(indexOp)) {
+                    logger.error("No Index operation specified for {}, transferring to failure.", new Object[]{file});
+                    flowFilesToTransfer.remove(file);
+                    file = session.penalize(file);
```

Comment:
Same comment as in Fetch, will penalizing this flowfile potentially help it be processed later? If not then it probably shouldn't be.